File: shared/origami/python/tests/test_ml_recommender.py
# Copyright Advanced Micro Devices, Inc., or its affiliates.
# SPDX-License-Identifier: MIT

"""Tests for the ML recommender tile prediction module.

Verifies layout classification, tile prediction, config ranking, and
weight loading via the origami.ml_recommender Python bindings.
"""

import pytest
import origami
from helpers import HARDWARE


# ---- Fixtures ----------------------------------------------------------------


@pytest.fixture
def hardware_gfx950():
    return HARDWARE["gfx950"]


def _make_bf16_problem(m, n, k, transA, transB, batch=1):
    """Helper to create a BFloat16 GEMM problem."""
    p = origami.problem_t()
    p.size = origami.dim3_t(m, n, k)
    p.batch = batch
    p.a_transpose = transA
    p.b_transpose = transB
    p.a_dtype = origami.data_type_t.BFloat16
    p.b_dtype = origami.data_type_t.BFloat16
    p.c_dtype = origami.data_type_t.BFloat16
    p.d_dtype = origami.data_type_t.BFloat16
    p.mi_dtype = origami.data_type_t.BFloat16
    return p


def _make_tf32_problem(m, n, k, transA, transB, batch=1):
    """Helper to create a TF32 (XFloat32) GEMM problem."""
    p = origami.problem_t()
    p.size = origami.dim3_t(m, n, k)
    p.batch = batch
    p.a_transpose = transA
    p.b_transpose = transB
    p.a_dtype = origami.data_type_t.Float
    p.b_dtype = origami.data_type_t.Float
    p.c_dtype = origami.data_type_t.Float
    p.d_dtype = origami.data_type_t.Float
    p.mi_dtype = origami.data_type_t.XFloat32
    return p


# ---- Layout classification --------------------------------------------------


class TestClassifyLayout:
    """Tests for origami.ml_recommender.classify_layout()."""

    def test_bbs_tn(self):
        p = _make_bf16_problem(1024, 2048, 512, origami.transpose_t.T, origami.transpose_t.N)
        assert origami.ml_recommender.classify_layout(p) == origami.ml_recommender.BBS_TN

    def test_bbs_nn(self):
        p = _make_bf16_problem(1024, 2048, 512, origami.transpose_t.N, origami.transpose_t.N)
        assert origami.ml_recommender.classify_layout(p) == origami.ml_recommender.BBS_NN

    def test_sssmx_nt(self):
        p = _make_tf32_problem(1024, 2048, 512, origami.transpose_t.N, origami.transpose_t.T)
        assert origami.ml_recommender.classify_layout(p) == origami.ml_recommender.SSS_MX_NT

    def test_half_maps_to_bbs_tn(self):
        p = _make_bf16_problem(1024, 2048, 512, origami.transpose_t.T, origami.transpose_t.N)
        p.a_dtype = origami.data_type_t.Half
        p.b_dtype = origami.data_type_t.Half
        assert origami.ml_recommender.classify_layout(p) == origami.ml_recommender.BBS_TN

    def test_fallback_for_unsupported_layout(self):
        """BBS_TT (both transposed) has no trained model; should fall back to BBS_NN."""
        p = _make_bf16_problem(1024, 2048, 512, origami.transpose_t.T, origami.transpose_t.T)
        assert origami.ml_recommender.classify_layout(p) == origami.ml_recommender.BBS_NN

    def test_f32_sgemm_falls_back(self):
        """Pure SGEMM (mi_dtype=Float) should not match SSS_MX_NT."""
        p = origami.problem_t()
        p.size = origami.dim3_t(1024, 2048, 512)
        p.batch = 1
        p.a_transpose = origami.transpose_t.N
        p.b_transpose = origami.transpose_t.T
        p.a_dtype = origami.data_type_t.Float
        p.b_dtype = origami.data_type_t.Float
        p.mi_dtype = origami.data_type_t.Float
        assert origami.ml_recommender.classify_layout(p) == origami.ml_recommender.BBS_NN


# ---- Tile prediction ---------------------------------------------------------


class TestPredictTile:
    """Tests for origami.ml_recommender.predict_tile()."""

    def test_returns_valid_dims(self):
        p = _make_bf16_problem(2048, 4096, 1024, origami.transpose_t.T, origami.transpose_t.N)
        tile = origami.ml_recommender.predict_tile(p)
        assert tile.mt_m > 0
        assert tile.mt_n > 0
        assert tile.mt_k > 0

    def test_deterministic(self):
        p = _make_bf16_problem(512, 1024, 256, origami.transpose_t.T, origami.transpose_t.N)
        t1 = origami.ml_recommender.predict_tile(p)
        t2 = origami.ml_recommender.predict_tile(p)
        assert t1.mt_m == t2.mt_m
        assert t1.mt_n == t2.mt_n
        assert t1.mt_k == t2.mt_k
        assert t1.score == pytest.approx(t2.score)

    @pytest.mark.skipif(
        not origami.ml_recommender.weights_loaded(),
        reason="Weights not loaded",
    )
    def test_different_shapes_different_tiles(self):
        small = _make_bf16_problem(32, 64, 128, origami.transpose_t.T, origami.transpose_t.N)
        large = _make_bf16_problem(16384, 16384, 4096, origami.transpose_t.T, origami.transpose_t.N)
        ts = origami.ml_recommender.predict_tile(small)
        tl = origami.ml_recommender.predict_tile(large)
        assert (ts.mt_m, ts.mt_n, ts.mt_k) != (tl.mt_m, tl.mt_n, tl.mt_k)

    @pytest.mark.skipif(
        not origami.ml_recommender.weights_loaded(),
        reason="Weights not loaded",
    )
    def test_batched_problem(self):
        p = _make_bf16_problem(64, 64, 128, origami.transpose_t.T, origami.transpose_t.N, batch=2048)
        tile = origami.ml_recommender.predict_tile(p)
        assert tile.mt_m > 0
        assert tile.score != pytest.approx(-1.0)


# ---- Config ranking ----------------------------------------------------------


class TestRankConfigs:
    """Tests for origami.ml_recommender.rank_configs()."""

    def _make_configs(self):
        configs = []
        for mt_m, mt_n, mt_k in [
            (256, 256, 64),
            (128, 128, 64),
            (192, 256, 64),
            (64, 64, 32),
            (128, 256, 32),
        ]:
            c = origami.config_t()
            c.mt = origami.dim3_t(mt_m, mt_n, mt_k)
            c.mi = origami.dim3_t(16, 16, 16)
            c.occupancy = 1
            configs.append(c)
        return configs

    def test_returns_results(self, hardware_gfx950):
        p = _make_bf16_problem(2048, 2048, 2048, origami.transpose_t.T, origami.transpose_t.N)
        configs = self._make_configs()
        ranked = origami.ml_recommender.rank_configs(p, hardware_gfx950, configs)
        assert len(ranked) > 0
        assert len(ranked) <= len(configs)

    def test_exact_match_ranked_first(self, hardware_gfx950):
        p = _make_bf16_problem(2048, 2048, 2048, origami.transpose_t.T, origami.transpose_t.N)
        predicted = origami.ml_recommender.predict_tile(p)

        configs = self._make_configs()
        exact = origami.config_t()
        exact.mt = origami.dim3_t(predicted.mt_m, predicted.mt_n, predicted.mt_k)
        exact.mi = origami.dim3_t(16, 16, 16)
        exact.occupancy = 1
        configs.append(exact)

        ranked = origami.ml_recommender.rank_configs(p, hardware_gfx950, configs)
        assert ranked[0].config.mt.m == predicted.mt_m
        assert ranked[0].config.mt.n == predicted.mt_n
        assert ranked[0].config.mt.k == predicted.mt_k

    def test_empty_configs(self, hardware_gfx950):
        p = _make_bf16_problem(1024, 1024, 1024, origami.transpose_t.T, origami.transpose_t.N)
        ranked = origami.ml_recommender.rank_configs(p, hardware_gfx950, [])
        assert len(ranked) == 0

    def test_stable_ordering(self, hardware_gfx950):
        p = _make_bf16_problem(4096, 4096, 512, origami.transpose_t.T, origami.transpose_t.N)
        configs = self._make_configs()
        r1 = origami.ml_recommender.rank_configs(p, hardware_gfx950, configs)
        r2 = origami.ml_recommender.rank_configs(p, hardware_gfx950, configs)
        assert len(r1) == len(r2)
        for a, b in zip(r1, r2):
            assert a.config.mt.m == b.config.mt.m
            assert a.config.mt.n == b.config.mt.n
            assert a.config.mt.k == b.config.mt.k


# ---- Weight loading ----------------------------------------------------------


class TestWeights:
    """Tests for weight loading functions."""

    def test_weights_loaded_returns_bool(self):
        result = origami.ml_recommender.weights_loaded()
        assert isinstance(result, bool)

    def test_load_nonexistent_returns_false(self):
        result = origami.ml_recommender.load_weights("/nonexistent/path.bin")
        assert result is False


# ---- prediction_modes_t enum ------------------------------------------------


class TestPredictionMode:
    """Verify ml_recommender is available in prediction_modes_t."""

    def test_ml_recommender_enum_exists(self):
        mode = origami.prediction_modes_t.ml_recommender
        assert mode is not None
        assert int(mode) == 2

File: shared/origami/tests/test_ml_recommender.cpp
/*******************************************************************************
 *
 * MIT License
 *
 * Copyright 2026 AMD ROCm(TM) Software
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 *******************************************************************************/

#include <catch2/catch_approx.hpp>
#include <catch2/catch_test_macros.hpp>
#include "common.hpp"
#include "origami/ml_recommender.hpp"

using Catch::Approx;
namespace ml = origami::ml_recommender;

// ---------------------------------------------------------------------------
// classify_layout
// ---------------------------------------------------------------------------

TEST_CASE("ML Recommender: classify_layout BBS_TN", "[ml_recommender]") {
  auto problem    = make_problem(1024, 2048, 512, origami::transpose_t::T, origami::transpose_t::N);
  problem.a_dtype = origami::data_type_t::BFloat16;
  problem.b_dtype = origami::data_type_t::BFloat16;
  auto key        = ml::classify_layout(problem);
  REQUIRE(key == ml::layout_key_t::BBS_TN);
}

TEST_CASE("ML Recommender: classify_layout BBS_NN", "[ml_recommender]") {
  auto problem    = make_problem(1024, 2048, 512, origami::transpose_t::N, origami::transpose_t::N);
  problem.a_dtype = origami::data_type_t::BFloat16;
  problem.b_dtype = origami::data_type_t::BFloat16;
  auto key        = ml::classify_layout(problem);
  REQUIRE(key == ml::layout_key_t::BBS_NN);
}

TEST_CASE("ML Recommender: classify_layout SSS_MX_NT", "[ml_recommender]") {
  auto problem    = make_problem(1024, 2048, 512, origami::transpose_t::N, origami::transpose_t::T);
  problem.a_dtype = origami::data_type_t::Float;
  problem.b_dtype = origami::data_type_t::Float;
  problem.mi_dtype = origami::data_type_t::XFloat32;
  auto key         = ml::classify_layout(problem);
  REQUIRE(key == ml::layout_key_t::SSS_MX_NT);
}

TEST_CASE("ML Recommender: classify_layout fallback to BBS_NN", "[ml_recommender]") {
  auto problem    = make_problem(1024, 2048, 512, origami::transpose_t::T, origami::transpose_t::T);
  problem.a_dtype = origami::data_type_t::BFloat16;
  problem.b_dtype = origami::data_type_t::BFloat16;
  auto key        = ml::classify_layout(problem);
  REQUIRE(key == ml::layout_key_t::BBS_NN);
}

TEST_CASE("ML Recommender: classify_layout Half maps to BBS_TN", "[ml_recommender]") {
  auto problem    = make_problem(1024, 2048, 512, origami::transpose_t::T, origami::transpose_t::N);
  problem.a_dtype = origami::data_type_t::Half;
  problem.b_dtype = origami::data_type_t::Half;
  auto key        = ml::classify_layout(problem);
  REQUIRE(key == ml::layout_key_t::BBS_TN);
}

// ---------------------------------------------------------------------------
// predict_tile (requires binary weights at runtime)
// ---------------------------------------------------------------------------

TEST_CASE("ML Recommender: predict_tile returns valid dimensions", "[ml_recommender]") {
  auto problem    = make_problem(2048, 4096, 1024);
  problem.a_dtype = origami::data_type_t::BFloat16;
  problem.b_dtype = origami::data_type_t::BFloat16;
  auto tile       = ml::predict_tile(problem);

  // Even without weights, predict_tile returns a safe default
  REQUIRE(tile.mt_m > 0);
  REQUIRE(tile.mt_n > 0);
  REQUIRE(tile.mt_k > 0);
}

TEST_CASE("ML Recommender: predict_tile deterministic", "[ml_recommender]") {
  auto problem    = make_problem(512, 1024, 256);
  problem.a_dtype = origami::data_type_t::BFloat16;
  problem.b_dtype = origami::data_type_t::BFloat16;

  auto tile1 = ml::predict_tile(problem);
  auto tile2 = ml::predict_tile(problem);

  REQUIRE(tile1.mt_m == tile2.mt_m);
  REQUIRE(tile1.mt_n == tile2.mt_n);
  REQUIRE(tile1.mt_k == tile2.mt_k);
  REQUIRE(tile1.score == Approx(tile2.score));
}

TEST_CASE("ML Recommender: predict_tile different shapes produce different tiles",
          "[ml_recommender]") {
  auto small_problem    = make_problem(32, 64, 128);
  small_problem.a_dtype = origami::data_type_t::BFloat16;
  small_problem.b_dtype = origami::data_type_t::BFloat16;

  auto large_problem    = make_problem(16384, 16384, 4096);
  large_problem.a_dtype = origami::data_type_t::BFloat16;
  large_problem.b_dtype = origami::data_type_t::BFloat16;

  auto tile_small = ml::predict_tile(small_problem);
  auto tile_large = ml::predict_tile(large_problem);

  if (ml::weights_loaded()) {
    bool different = (tile_small.mt_m != tile_large.mt_m) || (tile_small.mt_n != tile_large.mt_n) ||
                     (tile_small.mt_k != tile_large.mt_k);
    REQUIRE(different);
  }
}

// ---------------------------------------------------------------------------
// rank_configs
// ---------------------------------------------------------------------------

TEST_CASE("ML Recommender: rank_configs returns all valid configs", "[ml_recommender]") {
  auto hardware   = make_hardware(950);
  auto problem    = make_problem(2048, 2048, 2048);
  problem.a_dtype = origami::data_type_t::BFloat16;
  problem.b_dtype = origami::data_type_t::BFloat16;

  std::vector<origami::config_t> configs;
  configs.push_back(make_config(256, 256, 64));
  configs.push_back(make_config(128, 128, 64));
  configs.push_back(make_config(192, 256, 64));
  configs.push_back(make_config(64, 64, 32));

  auto ranked = ml::rank_configs(problem, hardware, configs);
  REQUIRE(ranked.size() > 0);
  REQUIRE(ranked.size() <= configs.size());
}

TEST_CASE("ML Recommender: rank_configs prefers exact tile match", "[ml_recommender]") {
  auto hardware   = make_hardware(950);
  auto problem    = make_problem(2048, 2048, 2048);
  problem.a_dtype = origami::data_type_t::BFloat16;
  problem.b_dtype = origami::data_type_t::BFloat16;

  auto predicted = ml::predict_tile(problem);

  // Use small tiles that always pass LDS capacity, plus the predicted tile
  std::vector<origami::config_t> configs;
  configs.push_back(make_config(64, 64, 32));
  configs.push_back(make_config(128, 128, 32));
  configs.push_back(
      make_config(predicted.mt_m, predicted.mt_n, predicted.mt_k));

  auto ranked = ml::rank_configs(problem, hardware, configs);
  REQUIRE(ranked.size() > 0);

  bool found_predicted = false;
  for (const auto& r : ranked) {
    if (r.config.mt.m == predicted.mt_m && r.config.mt.n == predicted.mt_n &&
        r.config.mt.k == predicted.mt_k) {
      found_predicted = true;
      break;
    }
  }
  REQUIRE(found_predicted);

  if (ranked[0].config.mt.m == predicted.mt_m &&
      ranked[0].config.mt.n == predicted.mt_n) {
    REQUIRE(ranked[0].config.mt.k == predicted.mt_k);
  }
}

TEST_CASE("ML Recommender: rank_configs handles empty configs", "[ml_recommender]") {
  auto hardware   = make_hardware(950);
  auto problem    = make_problem(1024, 1024, 1024);
  problem.a_dtype = origami::data_type_t::BFloat16;
  problem.b_dtype = origami::data_type_t::BFloat16;

  std::vector<origami::config_t> configs;
  auto ranked = ml::rank_configs(problem, hardware, configs);
  REQUIRE(ranked.empty());
}

TEST_CASE("ML Recommender: rank_configs stable across calls", "[ml_recommender]") {
  auto hardware   = make_hardware(950);
  auto problem    = make_problem(4096, 4096, 512);
  problem.a_dtype = origami::data_type_t::BFloat16;
  problem.b_dtype = origami::data_type_t::BFloat16;

  std::vector<origami::config_t> configs;
  configs.push_back(make_config(256, 256, 64));
  configs.push_back(make_config(128, 256, 64));
  configs.push_back(make_config(256, 128, 64));
  configs.push_back(make_config(192, 192, 64));

  auto ranked1 = ml::rank_configs(problem, hardware, configs);
  auto ranked2 = ml::rank_configs(problem, hardware, configs);

  REQUIRE(ranked1.size() == ranked2.size());
  for (std::size_t i = 0; i < ranked1.size(); ++i) {
    REQUIRE(ranked1[i].config.mt.m == ranked2[i].config.mt.m);
    REQUIRE(ranked1[i].config.mt.n == ranked2[i].config.mt.n);
    REQUIRE(ranked1[i].config.mt.k == ranked2[i].config.mt.k);
  }
}

// ---------------------------------------------------------------------------
// Multi-layout dispatch
// ---------------------------------------------------------------------------

TEST_CASE("ML Recommender: different layouts use different models", "[ml_recommender]") {
  auto tn_problem = make_problem(1024, 2048, 512, origami::transpose_t::T, origami::transpose_t::N);
  tn_problem.a_dtype = origami::data_type_t::BFloat16;
  tn_problem.b_dtype = origami::data_type_t::BFloat16;

  auto nn_problem = make_problem(1024, 2048, 512, origami::transpose_t::N, origami::transpose_t::N);
  nn_problem.a_dtype = origami::data_type_t::BFloat16;
  nn_problem.b_dtype = origami::data_type_t::BFloat16;

  auto tn_key = ml::classify_layout(tn_problem);
  auto nn_key = ml::classify_layout(nn_problem);
  REQUIRE(tn_key != nn_key);

  if (ml::weights_loaded()) {
    const auto& tn_w = ml::get_layout_weights(tn_key);
    const auto& nn_w = ml::get_layout_weights(nn_key);
    REQUIRE(tn_w.num_clusters > 0);
    REQUIRE(nn_w.num_clusters > 0);
  }
}

// ---------------------------------------------------------------------------
// weights_loaded / load_weights
// ---------------------------------------------------------------------------

TEST_CASE("ML Recommender: weights_loaded returns bool", "[ml_recommender]") {
  bool loaded = ml::weights_loaded();
  (void)loaded;
}
